File: dart/57.dart
class Solution {
    bool intersect(List<int> i, List<int> j) {
        if (i[0] > j[0])
            return intersect(j, i);
        return i[1] >= j[0];
    }

    int max(int x, int y){
        return (x > y) ? x : y;
    }

    int min(int x, int y){
        return (x < y) ? x : y;
    }

  List<List<int>> insert(List<List<int>> i, List<int> j) {
        int n = i.length, m = 0;
        bool d = false;
        for (int k = 0; k < n; k++) {
            if (intersect(i[k], j)) {
                int s = min(i[k][0], j[0]);
                while (k < n && i[k][0] <= j[1])
                    k++;
                i[m++] = [ s, max(i[--k][1], j[1]) ];
                d = true;
            } else
                i[m++] = i[k];
        }
        if (!d) {
            List<List<int>> r = [];
            for (int k = 0; k < n;) {
                if (!d && i[k][0] > j[0]) {
                    r.add(j);
                    d = true;
                } else
                    r.add(i[k++]);
            }
            if (!d)
                r.add(j);
            return r;
        }
        List<List<int>> r = [];
        for (int k = 0; k < m; k++)
            r.add(i[k]);
        return r;
  }
}

File: python3/57.py
class Solution:
    def insert(self, i: List[List[int]], j: List[int]) -> List[List[int]]:
        def intersect(i, j):
            if i[0] > j[0]:
                return intersect(j, i)
            return i[1] >= j[0]

        n = len(i)
        m = 0
        d = False
        k = 0
        while k < n:
            if intersect(i[k], j):
                s = min(i[k][0], j[0])
                while k < n and i[k][0] <= j[1]:
                    k += 1
                k -= 1
                i[m] = [s, max(i[k][1], j[1])]
                d = True
            else:
                i[m] = i[k]
            m += 1
            k += 1
        if not d:
            r = []
            k = 0
            while k < n:
                if (not d) and i[k][0] > j[0]:
                    r.append(j)
                    d = True
                else:
                    r.append(i[k])
                    k += 1
            if not d:
                r.append(j)
            return r
        print(i)
        return [i[k] for k in range(m)]
